File: projetos/cheatsheet/py/full.py
# MASTER CHEATSHEET PYTHON
# Tudo em um arquivo, comentários em português, seções separadas.
# Comentários explicativos curtos, exemplos práticos e outputs.

# =======================
# VARIÁVEIS
# =======================
# variáveis armazenam valores dinamicamente tipados
nome = "Bruno"
idade = 16
saldo = 100.50
ativo = True

print(nome)
#output
# Bruno

# atribuição múltipla
a, b, c = 1, 2, 3
print(a, b, c)
#output
# 1 2 3

# Variáveis podem mudar de tipo
x = 5
x = "agora string"
print(x)
#output
# agora string

# =======================
# TIPOS DE DADOS
# =======================
# principais tipos: str, int, float, bool, None, list, tuple, dict, set, bytes
s = "texto"
i = 10
f = 3.14
b = False
n = None
lista = [1, 2, 3]
tupla = (1, 2)
d = {"k": "v"}
conjunto = {1, 2}

print(type(s), type(i), type(f), type(b), type(n))
#output
# <class 'str'> <class 'int'> <class 'float'> <class 'bool'> <class 'NoneType'>

# conversões
print(int("5"), float("3.5"), str(10), bool(0))
#output
# 5 3.5 10 False

# =======================
# OPERADORES
# =======================
# Aritméticos: + - * / // % **
print(2 + 3, 5 - 2, 2 * 3, 7 / 2, 7 // 2, 7 % 2, 2 ** 3)
#output
# 5 3 6 3.5 3 1 8

# Atribuição: =, +=, -=, *=, /=, //=, **=
x = 5
x += 2
print(x)
#output
# 7

# Comparação: ==, !=, >, <, >=, <=
print(2 == 2, 2 != 3, 3 > 1)
#output
# True True True

# Lógicos: and, or, not
print(True and False, True or False, not True)
#output
# False True False

# Operador ternário (expressão condicional)
idade = 16
status = "maior" if idade >= 18 else "menor"
print(status)
#output
# menor

# None coalescing (operador padrão via or)
a = None
b = a or "valor padrao"
print(b)
#output
# valor padrao

# =======================
# FLUXO DE CONTROLE
# =======================
# if / elif / else
nota = 7
if nota >= 9:
    print("Excelente")
elif nota >= 6:
    print("Bom")
else:
    print("Precisa melhorar")
#output
# Bom

# match (Python 3.10+), similar a switch
valor = 2
match valor:
    case 1:
        print("um")
    case 2:
        print("dois")
    case _:
        print("outro")
#output
# dois

# break e continue em loops
for i in range(5):
    if i == 2:
        continue  # pula
    if i == 4:
        break     # sai
    print(i)
#output
# 0 1 3

# =======================
# LOOPS
# =======================
# for tradicional (iterável)
for i in range(3):
    print(i)
#output
# 0 1 2

# iterando lista
for item in ["a", "b", "c"]:
    print(item)
#output
# a b c

# while
c = 0
while c < 3:
    print(c)
    c += 1
#output
# 0 1 2

# loop com else
for i in range(3):
    print(i)
else:
    print("acabou")
#output
# 0 1 2 acabou

# =======================
# FUNÇÕES
# =======================
# declaração
def soma(a, b):
    return a + b

print(soma(2, 3))
#output
# 5

# parâmetros default
def cumprimenta(nome="amigo"):
    print("Olá", nome)

cumprimenta()
#output
# Olá amigo

# args / kwargs
def ex_args(*args, **kwargs):
    print("args:", args)
    print("kwargs:", kwargs)

ex_args(1, 2, a=3)
#output
# args: (1, 2)
# kwargs: {'a': 3}

# funções como objetos (passar como parâmetro)
def aplicar(fn, x):
    return fn(x)

print(aplicar(lambda v: v * 2, 5))
#output
# 10

# closures (função que "lembra" do ambiente)
def contador():
    count = 0
    def inc():
        nonlocal count
        count += 1
        return count
    return inc

c = contador()
print(c(), c())
#output
# 1 2

# funções de ordem superior (retornam/fazem funções)
def multiplicador(n):
    return lambda x: x * n

dobro = multiplicador(2)
print(dobro(5))
#output
# 10

# =======================
# LISTAS (arrays)
# =======================
# criar
l = [1, 2, 3]
vazia = []

# acessar
print(l[0], l[-1])
#output
# 1 3

# adicionar / remover
l.append(4)        # adiciona fim
l.insert(0, 0)     # insere índice
print(l.pop())     # remove último
l.remove(2)        # remove primeiro valor igual a 2

print(l)
#output
# 4 [0,1,3]

# slice (fatia) e cópia
arr = [1, 2, 3, 4]
fat = arr[1:3]     # [2,3]
copia = arr[:]     # copia
print(fat, copia)
#output
# [2, 3] [1, 2, 3, 4]

# concatenação
print([1, 2] + [3, 4])
#output
# [1, 2, 3, 4]

# comprehensions (essencial)
quad = [x * x for x in range(5)]
filt = [x for x in range(10) if x % 2 == 0]
print(quad, filt)
#output
# [0, 1, 4, 9, 16] [0, 2, 4, 6, 8]

# map / filter / reduce
from functools import reduce
print(list(map(lambda x: x * 2, [1, 2, 3])))
print(list(filter(lambda x: x > 1, [1, 2, 3])))
print(reduce(lambda a, b: a + b, [1, 2, 3]))
#output
# [2, 4, 6] [2, 3] 6

# enumerate e zip
for idx, val in enumerate(["a", "b"]):
    print(idx, val)
#output
# 0 a 1 b

for a, b in zip([1, 2], ["x", "y"]):
    print(a, b)
#output
# 1 x 2 y

# listas aninhadas e flatten simples
nested = [[1, 2], [3, 4]]
flat = [item for sub in nested for item in sub]
print(flat)
#output
# [1, 2, 3, 4]

# =======================
# TUPLAS
# =======================
t = (1, 2, 3)
print(t[0])
#output
# 1
# tuplas são imutáveis

# =======================
# DICIONÁRIOS (objects)
# =======================
d = {"nome": "Bruno", "idade": 16}
print(d["nome"])
#output
# Bruno

# métodos úteis
print(d.keys(), d.values(), d.items())
#output
# dict_keys(['nome', 'idade']) dict_values(['Bruno', 16]) dict_items([('nome', 'Bruno'), ('idade', 16)])

# adicionar/atualizar/remover
d["cidade"] = "Candelária"
d.update({"idade": 17})
print(d.pop("cidade"))
#output
# Candelária

# get com valor default
print(d.get("sobrenome", "NA"))
#output
# NA

# iteração
for k, v in d.items():
    print(k, v)
#output
# nome Bruno idade 17

# =======================
# SETS (conjuntos)
# =======================
s = {1, 2, 3}
s.add(4)
s.discard(2)
print(s)
#output
# {1, 3, 4}
# operações: union, intersection, difference

# =======================
# STRINGS
# =======================
txt = "  Olá Mundo  "
print(txt.strip())           # trim
print(txt.lower(), txt.upper())
print(txt.replace("Mundo", "Python"))
#output
# Olá Mundo
#   olá mundo    OLÁ MUNDO
#   Olá Python  

# f-strings (template literals)
nome = "Bruno"
print(f"Olá {nome}, {idade} anos")
#output
# Olá Bruno, 17 anos

# split / join
parts = "a,b,c".split(",")
print(parts, "-".join(parts))
#output
# ['a', 'b', 'c'] a-b-c

# =======================
# NÚMEROS & MATH
# =======================
import math
print(math.floor(4.7), math.ceil(4.1), round(4.5))
print(math.sqrt(16), math.pow(2, 3))
#output
# 4 5 4 4.0 8.0

import random
print(random.random(), random.randint(1, 10))
#output
# 0.123... 7

# =======================
# DATETIME
# =======================
from datetime import datetime, timedelta, date, time
agora = datetime.now()
print(agora)
#output
# 2025-xx-xx xx:xx:xx.xxxxxx

hoje = date.today()
amanha = hoje + timedelta(days=1)
print(hoje, amanha)
#output
# 2025-xx-xx 2025-xx-xx

# formatar
print(agora.strftime("%d/%m/%Y %H:%M:%S"))
#output
# 10/11/2025 17:00:00

# =======================
# ENTRADA E SAÍDA DE ARQUIVOS (automação essencial)
# =======================
# ler arquivo texto
with open("exemplo.txt", "w", encoding="utf-8") as f:
    f.write("linha1\nlinha2\n")

with open("exemplo.txt", "r", encoding="utf-8") as f:
    conteudo = f.read()
print(conteudo)
#output
# linha1
# linha2

# ler linha a linha
with open("exemplo.txt", "r", encoding="utf-8") as f:
    for linha in f:
        print(linha.strip())
#output
# linha1 linha2

# manipular caminhos com pathlib
from pathlib import Path
p = Path("exemplo.txt")
print(p.exists(), p.read_text(encoding="utf-8"))
#output
# True linha1\nlinha2\n

# copiar / mover / remover arquivos (shutil/os)
import shutil, os
shutil.copy("exemplo.txt", "exemplo_copy.txt")
os.remove("exemplo_copy.txt")

# CSV (importante para automação)
import csv
with open("dados.csv", "w", newline="", encoding="utf-8") as csvfile:
    writer = csv.writer(csvfile)
    writer.writerow(["nome", "idade"])
    writer.writerow(["Bruno", 16])

with open("dados.csv", "r", encoding="utf-8") as csvfile:
    reader = csv.DictReader(csvfile)
    for row in reader:
        print(row["nome"], row["idade"])
#output
# Bruno 16

# JSON (essencial)
import json
obj = {"a": 1}
s = json.dumps(obj)
print(s)
#output
# {"a": 1}
print(json.loads(s)["a"])
#output
# 1

# =======================
# REQUISIÇÕES HTTP (automação)
# =======================
# urllib padrão
import urllib.request, urllib.parse
url = "https://jsonplaceholder.typicode.com/todos/1"
with urllib.request.urlopen(url) as resp:
    data = resp.read().decode("utf-8")
print(data[:100])
#output
# {"userId": 1, "id": 1, ...

# requests (mais simples, requer pip install requests)
# import requests
# res = requests.get(url)
# print(res.json())

# =======================
# SUBPROCESS (executar comandos, automação)
# =======================
import subprocess
result = subprocess.run(["echo", "hello"], capture_output=True, text=True)
print(result.stdout.strip())
#output
# hello

# cuidado com segurança ao executar comandos externos

# =======================
# ERROS E EXCEÇÕES
# =======================
try:
    1 / 0
except ZeroDivisionError as e:
    print("erro:", e)
except Exception as e:
    print("erro generico:", e)
finally:
    print("sempre executa")
#output
# erro: division by zero
# sempre executa

# lançar exceções
def valida(n):
    if n < 0:
        raise ValueError("n deve ser >= 0")
    return n

# =======================
# MÓDULOS E PACOTES
# =======================
# criar/modularizar: criar arquivo .py e importar
# importar padrão: import math
# importar nome: from math import sqrt
# pacotes: pasta com __init__.py
# instalar pacotes: pip install nome_pacote

# =======================
# CLASSES E ORIENTAÇÃO A OBJETOS (essencial)
# =======================
class Pessoa:
    def __init__(self, nome, idade):
        self.nome = nome
        self.idade = idade

    def falar(self):
        print(f"Oi, sou {self.nome}")

p = Pessoa("Bruno", 16)
p.falar()
#output
# Oi, sou Bruno

# herança
class Aluno(Pessoa):
    def __init__(self, nome, idade, matricula):
        super().__init__(nome, idade)
        self.matricula = matricula

al = Aluno("A", 17, "123")
al.falar()
#output
# Oi, sou A

# =======================
# EXPRESSÕES LAMBDA E FUNÇÕES ANÔNIMAS
# =======================
f = lambda x, y: x + y
print(f(1, 2))
#output
# 3

# =======================
# COMPREHENSIONS AVANÇADAS
# =======================
d = {x: x * x for x in range(5)}
s = {x for x in range(5) if x % 2 == 0}
print(d, s)
#output
# {0: 0, 1: 1, 2: 4, 3: 9, 4: 16} {0, 2, 4}

# =======================
# REGEX BÁSICO
# =======================
import re
m = re.search(r"\d+", "abc123")
print(m.group())
#output
# 123

print(re.findall(r"\d+", "a1 b22 c333"))
#output
# ['1', '22', '333']

# =======================
# DEBUG E LOG
# =======================
print("mensagem")         # simples
import logging
logging.basicConfig(level=logging.INFO)
logging.info("info")
# para automação, prefira logging em vez de prints em scripts maiores

# =======================
# AMBIENTE E VARIÁVEIS
# =======================
# acessar variáveis de ambiente
print(os.environ.get("HOME", "NA"))
#output
# /home/usuario

# =======================
# UTILITÁRIOS IMPORTANTES PARA AUTOMAÇÃO
# =======================
# time.sleep
import time
time.sleep(0.1)

# shutil para copiar/pasta
# subprocess para rodar comandos
# pathlib para caminhos
# csv/json para dados tabulares e trocas
# urllib/requests para APIs

# =======================
# BOAS PRÁTICAS RÁPIDAS
# =======================
# - Use virtualenv/venv para projetos
# - Separe responsabilidades em módulos
# - Trate exceções previsíveis
# - Faça logging ao invés de prints em produção
# - Codifique em UTF-8 e explicit encoding quando ler/escrever arquivos
# - Evite executar comandos shell com dados não confiáveis

# =======================
# FIM DO CHEATSHEET PYTHON
# =======================
# Observação: arquivo focado em essenciais + automação.
# Pratique reescrevendo scripts que automatizem tarefas reais:
# leitura de CSV → transformação → gravar JSON / chamar API / mover arquivos.

File: projetos/cheatsheet/js/full.js
// MASTER CHEATSHEET JAVASCRIPT
// Tudo em um arquivo, comentários em português, seções separadas.
// Comentários explicativos, exemplos práticos e outputs curtos.

// =======================
// VARIÁVEIS
// =======================

// variáveis armazenam valores
// let - valor pode mudar
// const - valor fixo, nao pode reatribuir
// var - antiga, comportamento de escopo diferente (evitar)
let nome = "Bruno"
const idade = 16
var cidade = "Candelária"

console.log(nome)    // Bruno
console.log(idade)   // 16
console.log(cidade)  // Candelária

// scope: block, function, global
if (true) {
    let bloco = "so visivel aqui"
    // bloco existe apenas dentro do bloco
}
// function scope com var
function exemploScope() {
    var local = "visivel na funcao"
    console.log(local)
}
exemploScope()

// hoisting: declarações com var e funções são "içadas" antes da execução
hoistedFunction()
function hoistedFunction() {
    console.log("funcao hoisted")
}
// variaveis com var sao hoisted mas inicializadas com undefined
console.log(typeof hoistedVar) // undefined
var hoistedVar = 5

// =======================
// TIPOS DE DADO
// =======================

// String, Number, Boolean, Null, Undefined, Object, Symbol, BigInt
let str = "texto"
let num = 42
let bool = true
let vazio = null
let indefinido
let obj = { nome: "B" }
let sym = Symbol("id")
let big = 123456789012345678901234567890n

console.log(typeof str) // string
console.log(typeof num) // number
console.log(typeof bool) // boolean
console.log(typeof vazio) // object  (peculiaridade)
console.log(typeof indefinido) // undefined
console.log(typeof obj) // object
console.log(typeof sym) // symbol
console.log(typeof big) // bigint

// type checking com instanceof (para objetos criados por construtor)
const arr = [1,2,3]
console.log(arr instanceof Array) // true

// conversao de tipo
console.log(Number("123"))    // 123
console.log(String(123))      // "123"
console.log(Boolean(0))       // false

// =======================
// OPERADORES
// =======================

// Aritméticos: + - * / % **
console.log(2 + 3)   // 5
console.log(2 ** 3)  // 8

// Atribuição: = += -= *= /=
let a = 5
a += 2  // a = 7

// Comparação: == === != !== > < >= <=
console.log(2 == "2")  // true (valor)
console.log(2 === "2") // false (tipo+valor)

// Lógicos: && || !
console.log(true && false) // false
console.log(!false) // true

// Ternário
let pode = idade >= 18 ? "sim" : "nao"
console.log(pode)

// Nullish coalescing
let x = null
let y = x ?? "valor padrao"
console.log(y) // valor padrao

// Optional chaining
let usuario = { perfil: { nome: "B" } }
console.log(usuario?.perfil?.nome) // B
console.log(usuario?.endereco?.rua) // undefined

// =======================
// CONTROL FLOW
// =======================

// if / else if / else
let nota = 7
if (nota >= 9) {
    console.log("Excelente")
} else if (nota >= 6) {
    console.log("Bom")
} else {
    console.log("Precisa melhorar")
}

// switch
let dia = 3
switch (dia) {
    case 1:
        console.log("Segunda")
        break
    case 2:
        console.log("Terca")
        break
    default:
        console.log("Outro dia")
}

// break / continue
for (let i = 0; i < 5; i++) {
    if (i === 2) continue // pula o resto do loop
    if (i === 4) break    // sai do loop
    console.log(i)
}
// output: 0 1 3

// =======================
// LOOPS
// =======================

// for
for (let i = 0; i < 3; i++) {
    console.log(i)
}

// while
let v = 0
while (v < 3) {
    console.log(v)
    v++
}

// do while
let z = 0
do {
    console.log(z)
    z++
} while (z < 2)

// for...of (iterar valores de arrays)
for (const valor of [10,20,30]) {
    console.log(valor)
}

// for...in (iterar chaves de objetos)
const pessoa = { nome: "B", idade: 16 }
for (const chave in pessoa) {
    console.log(chave, pessoa[chave])
}

// =======================
// FUNÇÕES
// =======================

// function declaration
function soma(a, b) {
    return a + b
}
console.log(soma(2,3)) // 5

// function expression
const mult = function(a, b) {
    return a * b
}
console.log(mult(2,3)) // 6

// arrow function
const sub = (a, b) => a - b
console.log(sub(5,2)) // 3

// parametros padrao
function cumprimenta(nome = "amigo") {
    console.log(`Ola ${nome}`)
}
cumprimenta() // Ola amigo

// rest parameters
function total(...nums) {
    return nums.reduce((s, n) => s + n, 0)
}
console.log(total(1,2,3)) // 6

// callbacks
function executarCallback(cb) {
    cb("rodando")
}
executarCallback(msg => console.log(msg)) // rodando

// closures
function contador() {
    let count = 0
    return function() {
        count++
        return count
    }
}
const c = contador()
console.log(c()) // 1
console.log(c()) // 2

// higher-order functions: recebem ou retornam funções
function fazOperacao(fn, a, b) {
    return fn(a, b)
}
console.log(fazOperacao((x,y) => x*y, 3, 4)) // 12

// =======================
// ARRAYS
// =======================

// criar
let arr1 = [1,2,3]
let arr2 = new Array(3).fill(0) // [0,0,0]

// acessar por indice
console.log(arr1[0]) // 1
console.log(arr1.length) // 3

// push / pop
arr1.push(4) // adiciona no final
console.log(arr1.pop()) // remove e retorna 4

// shift / unshift
arr1.unshift(0) // adiciona no inicio
console.log(arr1.shift()) // remove inicio

// splice (mutável) e slice (não muta)
let a1 = [1,2,3,4]
a1.splice(1,2) // remove 2 itens a partir do indice 1 => a1 = [1,4]
let a2 = [1,2,3,4]
let parte = a2.slice(1,3) // [2,3]

// concat
let combinado = [1,2].concat([3,4]) // [1,2,3,4]
let combinado2 = [...[1,2], ...[3,4]] // spread

// join / split
let s = [1,2,3].join(",") // "1,2,3"
let fromString = "a,b,c".split(",") // ['a','b','c']

// map / filter / reduce / forEach
let numeros = [1,2,3]
console.log(numeros.map(x => x * 2)) // [2,4,6]
console.log(numeros.filter(x => x > 1)) // [2,3]
console.log(numeros.reduce((s,n) => s + n, 0)) // 6
numeros.forEach(x => console.log(x))

// find / findIndex / some / every
console.log(numeros.find(x => x === 2)) // 2
console.log(numeros.findIndex(x => x === 2)) // 1
console.log(numeros.some(x => x > 2)) // true
console.log(numeros.every(x => x > 0)) // true

// includes / indexOf / lastIndexOf
console.log(numeros.includes(2)) // true
console.log([1,2,2].lastIndexOf(2)) // 2

// sort / reverse
let txts = ["b","a","c"]
txts.sort() // altera o array
txts.reverse()

// flat / flatMap
let nested = [1, [2,3], [4, [5]]]
console.log(nested.flat()) // [1,2,3,4,[5]]
console.log([1,2].flatMap(x => [x, x*2])) // [1,2,2,4]

// spread e destructuring
let orig = [1,2,3]
let copia = [...orig] // copia
let [primeiro, segundo, ...resto] = orig

// copiar arrays
let copiaSlice = orig.slice()

// =======================
// OBJETOS
// =======================

// criar
let obj1 = { nome: "Bruno", idade: 16 }

// acessar
console.log(obj1.nome)
console.log(obj1["idade"])

// adicionar / atualizar / deletar
obj1.sobrenome = "Silva"
obj1.idade = 17
delete obj1.sobrenome

// Object.keys / values / entries
console.log(Object.keys(obj1))   // ['nome','idade']
console.log(Object.values(obj1)) // ['Bruno',17]
console.log(Object.entries(obj1)) // [['nome','Bruno'], ['idade',17]]

// spread e Object.assign
let objA = { a: 1 }
let objB = { b: 2 }
let combinadoObj = { ...objA, ...objB } // {a:1,b:2}
let alvo = {}
Object.assign(alvo, objA, objB)

// destructuring
const { nome: n, idade: i } = obj1

// metodos em objetos
let pessoa2 = {
    nome: "A",
    falar() {
        console.log("oi " + this.nome)
    }
}
pessoa2.falar()

// this keyword
const contexto = {
    nome: "X",
    getNome() {
        return this.nome
    }
}
console.log(contexto.getNome()) // X

// prototype e heranca (basico)
function Pessoa(nome) {
    this.nome = nome
}
Pessoa.prototype.falar = function() {
    console.log("Oi " + this.nome)
}
const p = new Pessoa("Proto")
p.falar()

// =======================
// STRINGS
// =======================

let texto = "  Ola Mundo  "

// template literals
let saud = `Ola ${nome}`

// slice / substring / substr
console.log(texto.slice(2,5))
console.log(texto.substring(2,5))

// replace / replaceAll
console.log("foo bar foo".replace("foo","baz"))
console.log("foo foo".replaceAll("foo","x"))

// split / join
console.log("a,b,c".split(",")) // ['a','b','c']
console.log(["a","b"].join("-")) // "a-b"

// trim / toLowerCase / toUpperCase
console.log(texto.trim())
console.log(texto.toLowerCase())
console.log(texto.toUpperCase())

// includes / startsWith / endsWith
console.log("hello".includes("ell")) // true
console.log("hello".startsWith("he")) // true

// charAt / indexOf / lastIndexOf
console.log("abc".charAt(1)) // b
console.log("abcabc".indexOf("b")) // 1

// padStart / padEnd / repeat
console.log("5".padStart(3,"0")) // "005"
console.log("x".repeat(3)) // "xxx"

// =======================
// NUMBERS
// =======================

console.log(Math.floor(4.7)) // 4
console.log(Math.ceil(4.1)) // 5
console.log(Math.round(4.5)) // 5
console.log(Math.random())   // 0..1
console.log(Math.min(1,2,3))
console.log(Math.max(1,2,3))

console.log(parseInt("10")) // 10
console.log(parseFloat("3.14")) // 3.14

console.log((2/3).toFixed(2)) // "0.67"
console.log((123.456).toPrecision(4)) // "123.5"

// =======================
// DATE & TIME
// =======================

const agora = new Date()
console.log(agora.getDate()) // dia do mes
console.log(agora.getMonth()) // mes (0-11)
console.log(agora.getFullYear())
console.log(agora.getHours(), agora.getMinutes(), agora.getSeconds())
console.log(agora.toLocaleString())
console.log(agora.toISOString())

// =======================
// JSON
// =======================

const objJson = { a: 1 }
const strJson = JSON.stringify(objJson) // '{"a":1}'
console.log(JSON.parse(strJson).a) // 1

// =======================
// TRATAMENTO DE ERROS
// =======================

try {
    throw new Error("erro custom")
} catch (err) {
    console.error("pegou erro:", err.message)
} finally {
    console.log("sempre executa")
}

// =======================
// JAVASCRIPT ASSINCRONO
// =======================

// setTimeout / setInterval
setTimeout(() => console.log("delay 1s"), 1000)
const id = setInterval(() => console.log("interval 1s"), 1000)
clearInterval(id) // parar

// Promises
const promessa = new Promise((resolve, reject) => {
    setTimeout(() => resolve("feito"), 100)
})
promessa.then(res => console.log(res)).catch(err => console.error(err)).finally(() => {})

// async / await
async function buscar() {
    try {
        const res = await promessa
        console.log("await:", res)
    } catch (e) {
        console.error(e)
    }
}
buscar()

// fetch (exemplo basico, funciona no navegador ou com node + fetch)
async function buscaApi(url) {
    try {
        const res = await fetch(url)
        const data = await res.json()
        console.log(data)
    } catch (err) {
        console.error("erro fetch", err)
    }
}
// exemplo: buscaApi('https://jsonplaceholder.typicode.com/todos/1')

// =======================
// DOM MANIPULATION (NAVEGADOR)
// =======================

// selecionar
// document.querySelector('.classe')
// document.querySelectorAll('div')
// document.getElementById('id')
// document.getElementsByClassName('classe')

// editar texto / html
// const el = document.querySelector('h1')
// el.textContent = 'Novo texto'
// el.innerHTML = '<span>html</span>'

// editar estilo
// el.style.color = 'red'
// el.classList.add('ativo')

// criar / remover elementos
// const novo = document.createElement('div')
// novo.textContent = 'oi'
// document.body.append(novo)
// novo.remove()

// eventos
// el.addEventListener('click', e => console.log(e.target))
// el.removeEventListener('click', handler)

// propagacao: bubbling e capturing
// e.stopPropagation() para parar bolha

// =======================
// MÓDULOS
// =======================

// export / import (ES modules)
// export function fn() {}
// import { fn } from './arquivo.js'
// default vs named exports
// export default function() {}
// import nome from './arquivo.js'

// organização: separar responsabilidades em arquivos e importar

// =======================
// CONCEITOS AVANCADOS
// =======================

// hoisting: var e function sao içadas
// scope chain: busca de variaveis caminha pelos scopes
// this binding: depende de como a funcao foi chamada
const exemploThis = {
    nome: "X",
    falar: function() { console.log(this.nome) }
}
exemploThis.falar() // X

// call, apply, bind
function saudacao(sufixo) {
    console.log(this.nome + sufixo)
}
const ctx = { nome: "Ctx" }
saudacao.call(ctx, "!")   // Ctx!
saudacao.apply(ctx, ["?"])// Ctx?
const bindado = saudacao.bind(ctx, "!!")
bindado() // Ctx!!

// closures: revisar para encapsular estado e criar funções privadas

// prototype chain: heranca via prototipo
// classes: sintaxe moderna para construtores
class Animal {
    constructor(nome) {
        this.nome = nome
    }
    falar() {
        console.log(this.nome + " faz som")
    }
}
class Cachorro extends Animal {
    falar() {
        console.log(this.nome + " late")
    }
}
const dog = new Cachorro("Rex")
dog.falar()

// destructuring aninhado
const dados = { user: { id: 1, profile: { idade: 16 } } }
const { user: { profile: { idade: idadeUser } } } = dados
console.log(idadeUser) // 16

// pure vs impure functions
// pure: mesma entrada = mesma saida, sem efeitos colaterais
// impure: depende de ou altera estado externo

// =======================
// UTILITÁRIOS
// =======================

// localStorage / sessionStorage (navegador)
// localStorage.setItem('chave', 'valor')
// localStorage.getItem('chave')

// debugging
console.log("log")
console.table([{a:1},{a:2}])
console.error("erro")

// regex basico
const re = /ab+c/
console.log(re.test("abbb")) // true
console.log("abc123".match(/\d+/)) // ['123']

// =======================
// FIM DO CHEATSHEET
// =======================

// Observacao: esse arquivo é um manual compacto.
// Em projetos reais modularize o codigo e evite poluir o escopo global.
// Pratique reescrevendo e recriando pequenos projetos com essas tecnicas.
